### DISCUSSION TITLE
Support tabs by optionally displaying multiple windows in a column as tabs

### DISCUSSION THREAD

**Discussion by elkowar** [YaLTeR/niri#876] [Ideas]
## Concept 

This is a simple concept for how to add tabs into niri, without conflicting with any of the other potential ideas for future 2D scrolling, etc, and without requiring an additional set of navigation keybinds.
Instead of adding tabs as a completely separate feature, the idea is to allow any columns (vertically stacked windows) to alternatively be displayed as a set of tabs, rather than a vertical stack of windows.

![image](https://github.com/user-attachments/assets/b676907d-15a8-42ca-8ede-8f42033ee725)

While this still brings some limitations (you couldn't have a tab group be horizontally split), this, in return, solves most of the issues regarding the fact that tabs require a completely new set of keybinds, have complex and often unclear interactions with the rest of the UX, etc.
The semantics of the keybinds from columns would carry over 1:1, as the tab list could be displayed vertically.

This would require one additional keybind, to toggle a column between those two states, as well as one additional configuration option to define the default mode for columns.

## Alternatives

There have been many different ideas floating around, such as allowing each column to scroll vertically (which would serve the same usecase, but with different UX), or adding a more general 2D scrolling concept.

Additionally, some of the usecases for tabbing could be served by having a "sticky".

All of these are much larger design questions, probably more complex to implement, and have way more difficult to answer questions in regard to how they interact with the rest of niri.

An additional variant of this would be to just resize the windows, which you can technically already hack together a bad version of via `is-active-in-column` and min/max-height in a window rule:
![image](https://github.com/user-attachments/assets/3923f03f-471f-48a3-8af5-bbe03c360708)

**Comment by YaLTeR** [comment]
Yeah, would be very interesting to try this. Probably not even that difficult to try (though without window titles because we don't have text rendering hooked up well enough).

**Reply by elkowar** [reply]
🤷🏻‍♂️ I'd love to have this even without window titles, definitely.

**Reply by YaLTeR** [reply]
Hm, how should this work visually when using CSD and no border/focus ring? Especially since many windows have top or all rounded corners in this case, and we likely don't know the radii.

**Comment by bbb651** [comment]
> An additional variant of this would be to just resize the windows, which you can technically already hack together a bad version of via is-active-in-column and min/max-height in a window rule:

To clarify, my idea was not to just resize the windows, but also "collapse" them so they only contain the window title. So from a user perspective they *are* vertical tabs (niri could represent them either as a column with collapsed windows or as a special tab container), with the benefit of reducing wasted vertical space that the active windows's tab would take (imo this is the main downside of vertical tabs, in both versions). I'm not sure how the UX of this would feel tough, and if the alternative way to implement it as special tiles is better.

**Reply by elkowar** [reply]
I guess this could just be a separate design for the tab bar, where instead of showing a typical horizontal tab bar, you just show the tabs "above" the active window above the window, and the tabs "below" the active window below.

**Comment by salman-farooq-sh** [comment]
How much work is vertically scrolling windows within a column?

Is it unanimously agreed that they are the best solution for this (disregarding the amount of work needed) or are there any benefits unique to other approaches like this?

They are much more nicer and are worth waiting for in my opinion.

**Reply by YaLTeR** [reply]
Hm, how is vertical scrolling within a column related here?

**Reply by salman-farooq-sh** [reply]
Both of:

1. tabbed columns as described here, and
2. vertical scrolling within columns on overflow i.e. columns also become a scrolling strip when the sum of heights of windows in a column is more than screen height

...provide a way to quickly swap what window a column is showing without changing what is shown in the columns on the left and right side.

#2 is more flexible as it can show two half height windows at the same time too instead of being bound to show only a single full height window at a time. No need for new shortcuts too because it is intuitive. Regarding gestures... maybe use a modifier or a normal key or different amount of fingers for scrolling within the column strip with up/down gesture instead of scrolling to other desktops up and below.

I think I made an issue for this previously. Let me find and attach it here.

**Reply by salman-farooq-sh** [reply]
https://github.com/YaLTeR/niri/issues/514

This one, a bit outdated though because I think at that time the surplus windows were just clipped out.

**Reply by elkowar** [reply]
I do think vertical scrolling is a relevant alternative to consider here. However, I'd argue that the """best""" approach might be to support two or three different modes for a column to be in:
- tabbed
- split  (the way it is right now)
- scrolling (vertical scrolling, which technically could just be a less constrained version of split)

They all allow different ways to show multiple windows in one column, and they all have more or less identical semantics when it comes to keybinds and most other things. So, from a design point of view, I see little reason not to want to have all three options!

Of course, implementation complexity is a different question.
Additionally, I do think that vertical scrolling columns might interact more directly with other design concepts regarding 2D navigation that have been thrown around, so I think from a pure perspective of "how likely is it that this feature might cause problems in the future", tabs might be a bit less problematic --that's purely by intuition though, maybe i'm wrong here.

**Reply by YaLTeR** [reply]
I'd start with tabs as a second mode to the current split. I feel vertical scrolling would be more effort and with more potential impl problems.

**Reply by salman-farooq-sh** [reply]
I think fully free flowing 2d navigation i.e. as in an infinite 2d canvas where windows can be put anywhere... it changes niri completely. It loses the concept of workspaces or at least makes every workspace a separate such 2d canvas (gestures how?). So we are getting a third dimension here too.

2d navigation would imply diagonal scrolling? I think that can be a bit unwieldy because it can get difficult to precisely switch back and forth between two windows. And if no diagonal navigation then it is better to have vertically scrolling columns as otherwise you reach at what niri already is i.e. vertically scroll for workspaces and horizontally on a strip.

As splitting a column is already present/done, we only need one more of tabs or column-scroll to solve the the real problem. Doing two things that essentially solve the same problem is unnecessary effort.

And my vote is for column-scroll, as it not only looks prettier with the scrolling animations which are already bread-and-butter of niri but it will also not need new design for any variation of tabs which allows for a more minimal looking setup because less UI elements.

Nice animations can be done for tabs too I am sure, but will probably be more effort to make them feel at home with rest of niri.

**Reply by salman-farooq-sh** [reply]
Is there any github discussion or unique term we are using (because we are already 2d technically) for 2d navigation as it is discussed here?

**Reply by YaLTeR** [reply]
2d is quite different from niri's current independent horizontal strips. Other projects like newm explore that concept

**Reply by salman-farooq-sh** [reply]
Cool! So will we ever do newm-style 2d? or is it firmly forever out of scope for niri?

**Reply by YaLTeR** [reply]
Sounds pretty out of scope to me tbh

**Reply by salman-farooq-sh** [reply]
I agree.

**Reply by salman-farooq-sh** [reply]
So coming to my original comment, column tabs before column scrolling only because of amount of effort? I completely understand if so because to solve the problem faster. If you or someone else decides to add column scrolling later down the line, will both of these co-exist then or you only want maintenance burden of one? because the last thing we want is niri to get too difficult to be maintained and abandoned consequently.

**Reply by YaLTeR** [reply]
I like the tabs idea

**Comment by YaLTeR** [comment]
Opened an issue to track this: https://github.com/YaLTeR/niri/issues/933

**Comment by YaLTeR** [comment]
I have a draft implementation ready at https://github.com/YaLTeR/niri/pull/1085, looking for feedback.
